[WK2] Drop use of fixed-width types for size values in ArgumentCoder specializations

zdobersek opened the pull request:
#### d5514afba482699970de2ac435d31bb1c7acfc0b
<pre>
[WK2] Drop use of fixed-width types for size values in ArgumentCoder specializations
<a href="https://bugs.webkit.org/show_bug.cgi?id=250777">https://bugs.webkit.org/show_bug.cgi?id=250777</a>

Reviewed by Kimmo Kinnunen.

Instead of finding better or worse fixed-sized-type matches for size values in
different ArgumentCoder specializations, the types of those size values should
be used. This means simply encoding values of size_t or unsigned types.

The target type of the size value should be noted in the code, just to provide
the necessary context of what is being encoded and have it match the decoding
side. This means the size type is explicitly used either in any temporary
variable where the value is stored, or an additional static_cast is used to
explicitly specify the type when the value is passed to the stream insertion
operator.

This still leaves open the possibility of using mismatched types, i.e. the
size value type not matching what the encoding enforces. Some sort of validation
could be implemented in the encoder class to make sure no undesired narrowing
conversions occur.

* Source/WebKit/Platform/IPC/ArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;CString&gt;::encode):
(IPC::ArgumentCoder&lt;CString&gt;::decode):
(IPC::ArgumentCoder&lt;String&gt;::encode):
(IPC::decodeStringText):
(IPC::ArgumentCoder&lt;String&gt;::decode):
(IPC::ArgumentCoder&lt;StringView&gt;::encode):
* Source/WebKit/Platform/IPC/ArgumentCoders.h:

Canonical link: <a href="https://commits.webkit.org/259488@main">https://commits.webkit.org/259488@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e44e2be5e91be15b0c858c05dc005c69c7709a50

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104411 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13489 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37320 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113627 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14591 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4406 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96637 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112665 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110179 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11248 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94308 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38871 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93112 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25919 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80540 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/state-changed (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6850 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27276 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6979 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3829 "Found 2 new test failures: fast/images/avif-as-image.html, fast/images/avif-heif-container-as-image.html (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13006 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46834 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6544 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8773 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebKit/Platform/IPC/ArgumentCoders.cpp:
```diff
@@ -41,23 +41,23 @@ void ArgumentCoder<CString>::encode(Encoder& encoder, const CString& string)
     }
 
-    encoder << static_cast<uint32_t>(string.length());
+    encoder << validateType<size_t>(string.length());
```
`encode.template encode<size_t>(string.length());`

Which is more appropriate? I don't know..

zdobersek replied:
Idea was to have compile-time validation of what the expected type returned by the size accessor is. This approach does it through a templated pass-through, avoiding adding static asserts everywhere.

Along with compile-time validation, it also specifies clearly the type that's going into the encoder, giving clear indication of the type that's expected on the decoding side. Just extra assurance against any future changes that might unknowingly end up tweaking these parts.

kkinnunen-apple replied:
Yeah, that part is good. 
What I mean is that there's an option:
Either:
a) validate via `encoder << validateType<size_t>(...)`
Or:
b) enforce via `encode.template encode<size_t>(...)`
Or:
c) enforce via `encoder << static_cast<size_t>(..)`


So the question is why validate instead of enforce?

If the enforce would be missing, in that case the validate would be missing aswell, so there's no additional security guarantee ? 
Maybe the benefit of validate would be if there are warnings of redundant static_casts?
Or that we think that a developer will remove the redundant static cast?

zdobersek replied:
Validating, as done here, forces adjustment to the type that's returned by the size accessor, no unwitting conversion to anything else allowed. It's a bit clunky to work with, and as seen it doesn't really cause anything to be fixed at the moment.

Enforcing can, on integer types, cause problematic conversions. But at the moment I think that's not happening in any of these specializations, so I would just switch to that. b) seems nicest, and it allows to eventually implement validation in a more refined way.

zdobersek replied:
I thought `encoder.template encoder<T>()` exists, but it doesn't. Can be added later. Going with c).

zdobersek committed:
[WK2] Drop use of fixed-width types for size values in ArgumentCoder specializations
https://bugs.webkit.org/show_bug.cgi?id=250777

Reviewed by Kimmo Kinnunen.

Instead of finding better or worse fixed-sized-type matches for size values in
different ArgumentCoder specializations, the types of those size values should
be used. This means simply encoding values of size_t or unsigned types.

The target type of the size value should be noted in the code, just to provide
the necessary context of what is being encoded and have it match the decoding
side. This means the size type is explicitly used either in any temporary
variable where the value is stored, or an additional static_cast is used to
explicitly specify the type when the value is passed to the stream insertion
operator.

This still leaves open the possibility of using mismatched types, i.e. the
size value type not matching what the encoding enforces. Some sort of validation
could be implemented in the encoder class to make sure no undesired narrowing
conversions occur.

* Source/WebKit/Platform/IPC/ArgumentCoders.cpp:
(IPC::ArgumentCoder<CString>::encode):
(IPC::ArgumentCoder<CString>::decode):
(IPC::ArgumentCoder<String>::encode):
(IPC::decodeStringText):
(IPC::ArgumentCoder<String>::decode):
(IPC::ArgumentCoder<StringView>::encode):
* Source/WebKit/Platform/IPC/ArgumentCoders.h:

Canonical link: https://commits.webkit.org/259488@main
